### DISCUSSION TITLE
DOC: uniqueness in ex05

### DISCUSSION THREAD

**Discussion by gdmcbain** [kinnala/scikit-fem#751] [General]
Not urgent.  (And possibly placed somewhere better than in this issue tracker?  Back on the Computational Science Stack Exchange?)

ex05 purports to solve Δ _u_ = 0 in the unit square subject to homogeneous
* Dirichlet  _u_ = 0 on _y_ = 0 and
* Neumann **n** ⋅ ∇ _u_ = 0 on _x_ = 0 and  _y_ = 1

conditions on three sides but driven by an integrally constrained flux on the fourth: ∫ **n** ⋅ ∇ _u_  d _y_ =  1, with the integral being over _x_ = 1.

However, this problem, posed like this, does not have a unique solution; _u_ [ _n_]  =  sin {(_n_ + ½)π _y_} cosh {(_n_ + ½)π _x_} / sinh {(_n_ + ½)π} for _n_ ∈ ℕ is a solution, as is any linear combination with weights summing to unity.  Whereas this kind of nonuniqueness often results in a singular finite element stiffness matrix (e.g. the [pure Neumann problem](https://dealii.org/developer/doxygen/deal.II/step_11.html)), this is not the case here.  At first glance, the numerical solution appears to resemble _u_ [0].

This kind of integral constraint is discussed at:
* [Solving Poisson equation with current BC using FEM](https://scicomp.stackexchange.com/questions/26967/solving-poisson-equation-with-current-bc-using-fem)
* [How to apply an integrated constrain condition in FEM?](https://scicomp.stackexchange.com/questions/28562/how-to-apply-an-integrated-constrain-condition-in-fem#comment51786_28562)

The discussions imply that this is a reasonably standard boundary condition (‘I know commercial software solves this problem just as efficiently as with regular BC’) and don't mention uniqueness.

Questions: 
* What is the mathematical status of this boundary condition?
* Why doesn't the finite element discretization reflect the nonuniqueness?
* Why does the discrete problem have the particular unique solution that it does?

Hunch: the discrete formulation is minimizing something which eliminates all _n_ > 0.

**Comment by gdmcbain** [comment]
> At first glance, the numerical solution appears to resemble u [0].

Not quite!  It is qualitatively similar but the top-right value (_x, y_) = (1, 1) of the numerical solution is 0.9357…  For the _n_-th exact solution, it is sin (_n_ + ½)π / tanh (_n_ + ½)π; e.g. for _n_ = 0, coth π/2 = 1.09033…

Some plots:

![ex05_solution](https://user-images.githubusercontent.com/1588947/55046856-f8bd9000-5096-11e9-88f9-60ffb366db1d.png)

_Figure:—_ The numerical solution, `ex05.x`.

![ex05-exact-0](https://user-images.githubusercontent.com/1588947/55046854-f824f980-5096-11e9-9454-bcb0b7b1dc94.png)
![ex05-exact-1](https://user-images.githubusercontent.com/1588947/55046852-f824f980-5096-11e9-9c80-21f6372a87f2.png)
![ex05-exact-2](https://user-images.githubusercontent.com/1588947/55046850-f78c6300-5096-11e9-9317-def7a0643e6e.png)

_Figure:—_ The first three exact solutions, _n_ = 0, 1, 2.

**Comment by gdmcbain** [comment]
I notice that `ex05.K` isn't symmetric, because `ex05.B` isn't.  Should the integral-flux condition break symmetry?

**Comment by gdmcbain** [comment]
A bit stumped after the above, I began to wonder what this kind of boundary condition might mean. One of the posters on scicomp referred to it as an ‘electrode’  through which a given current passes. Could the implication be that the electrode is at an unknown uniform potential?  Does that render the problem well posed?

**Reply by gatling-nrl** [reply]
I'm very interested in this. Did you have any more thoughts on whether the uniform potential constraint was adequate?

**Reply by gdmcbain** [reply]
'Uniform potential constraint '? Sure, but that's not in ex05. Yes, I can imagine modelling an electrode along a boundary with two constraints: 
1. electrode is isopotential
2. flux through electrode matches specified value

That seems well-posed and would be easy enough to do in scikit-fem. See multipoint-constraint #438 for the isopotential (not in ex05) and use a scalar Lagrange multiplier for the flux as already in ex05.

https://github.com/kinnala/scikit-fem/blob/71eb03e9baee4deab12c9fd91a04118bde3838ec/docs/examples/ex05.py#L54

I remain puzzled by what physical problem the current ex05 solves.

**Comment by kinnala** [comment]
Currently I don't have enough time to work on this. But let's leave the issue open so maybe someone with more knowledge on the subject pops up.

**Comment by gdmcbain** [comment]
O. K. 

I think I see three subproblems:
* The original problem of the uniqueness and what problem the current implementation is actually solving.  This one I find to be quite a puzzle.
* What the original poster at scicomp was after and what the ‘commercial software’ does.  I might ask some colleagues about that; when I first started looking at problems like ex13 in the industrial context, they were often solved with commerical software specifying a ‘current’ at a port.  That _might_ be worth an example, marginal though, in my opinion.
* Other kinds of integral constraints.  The classic one is the Neumann problem, rendered unique by specifying that the solution have zero mean (e.g. Braess 2001 _Finite Elements_, 2nd ed., Problem III.4.21).  That's probably worth an example and I expect would be implemented similarly to ex05.  An interesting iterative alternative is outlined by Elman, Silvester, & Wathen 2014 _Finite Elements and Fast Iterative Solvers_, 2nd ed., §2.3 ‘Singular systems are not a problem’).

**Comment by kinnala** [comment]
I thought about this for a while. I wonder if this particular strong formulation is an incorrect way of looking at this. This is after all a saddle point problem so I'd think the correct strong formulation would have two unknowns, u and some Lagrange multiplier. Also, I think not all u[n] are equivalent in that they have different energies (H^1 seminorm of the solution). u[0] corresponds to the smallest energy.
